generator/validate/gen.go

```diff
@@ -0,0 +1,178 @@
+		typeToCheck.Name,
+		unionDiscriminatorField.Name), unionDiscriminatorField.RawField)
+
+	discriminatorTypeRef, typeFound := root.TypesInfo.Types[unionDiscriminatorField.RawField.Type]
+	if !typeFound {
+		root.AddError(wrongTypeError)
+		return
+	}
+
+	underlyingType, underlyingTypeIsBasic := discriminatorTypeRef.Type.Underlying().(*types.Basic)
+	if !underlyingTypeIsBasic || underlyingType.Kind() != types.String {
+		root.AddError(wrongTypeError)
+		return
+	}
+
+	getEnumMarker := func(markers markers.MarkerValues) *crdmarkers.Enum {
+		enumMarkerIf := markers.Get("kubebuilder:validation:Enum")
+		if enumMarkerIf == nil {
+			return nil
+		}
+		enumMarker := enumMarkerIf.(crdmarkers.Enum)
+		return &enumMarker
+	}
+
+	var fieldEnumMarkerPtr *crdmarkers.Enum = getEnumMarker(unionDiscriminatorField.Markers)
+	var typeEnumMarkerPtr *crdmarkers.Enum = nil
+	var enumTypeInfo *markers.TypeInfo = nil
+
+	if enumTypeNamed, isNamed := discriminatorTypeRef.Type.(*types.Named); isNamed {
+		enumTypeInfo = packageTypes[enumTypeNamed.Obj().Name()]
+		if enumTypeInfo != nil {
+			typeEnumMarkerPtr = getEnumMarker(enumTypeInfo.Markers)
+		} else {
+			root.AddError(loader.ErrFromNode(fmt.Errorf(
+				"In union `%v` the union discriminator `%v` should have a string-based type defined in the same package",
+				typeToCheck.Name,
+				unionDiscriminatorField.Name), unionDiscriminatorField.RawField))
+		}
+	}
+
+	if fieldEnumMarkerPtr != nil && typeEnumMarkerPtr != nil {
+		root.AddError(loader.ErrFromNode(fmt.Errorf(
+			"Type `%v` should not define a `kubebuilder:validation:Enum` annotation, since it is already defined in the discriminator `%v` of union `%v`",
+			enumTypeInfo.Name,
+			unionDiscriminatorField.Name,
+			typeToCheck.Name), enumTypeInfo.RawSpec))
+	}
+
+	if fieldEnumMarkerPtr == nil && typeEnumMarkerPtr == nil {
+		root.AddError(loader.ErrFromNode(fmt.Errorf(
+			"In union `%v` the union discriminator `%v` should specify the allowed union values through a comment marker (kubebuilder:validation:Enum=....)",
+			typeToCheck.Name,
+			unionDiscriminatorField.Name), unionDiscriminatorField.RawField))
+		return
+	}
+
+	checkEnumValues := func(enumMarkerPtr *crdmarkers.Enum, node ast.Node) {
+		for index, enumValue := range *enumMarkerPtr {
+			stringValue, isString := enumValue.(string)
+			if !isString || stringValue != unionMembers[index] {
+				root.AddError(loader.ErrFromNode(fmt.Errorf(
+					"In union `%v` the union discriminator `%v` should allow the following values through the `kubebuilder:validation:Enum` comment marker: "+strings.Join(unionMembers, ";"),
+					typeToCheck.Name,
+					unionDiscriminatorField.Name), node))
+				break
+			}
+		}
+	}
+	if fieldEnumMarkerPtr != nil {
+		checkEnumValues(fieldEnumMarkerPtr, unionDiscriminatorField.RawField)
+	} else {
+		checkEnumValues(typeEnumMarkerPtr, enumTypeInfo.RawSpec)
```

Comment:
enumTypeInfo is nullable here

Reply:
Same answer here : `typeEnumMarkerPtr != nil` => `enumTypeInfo != nil`